### PR TITLE
Feature/table

### PR DESCRIPTION
**Pull request by allanhortle**
image

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "lint": "eslint src",
     "lint-fix": "eslint src --fix",
     "ava": "ava",
-    "test": "nyc --report-dir=./coverage npm run ava && npm run check-coverage",
+    "test": "nyc --report-dir=./coverage ava && npm run check-coverage",
```
Comment (dancoates) [inline package.json:16]: Not really part of this change but should this be `yarn run check-coverage`?

File: src/typedef/Typedefs.js [hunk 2]
```diff
@@ -7,4 +8,11 @@
  * @return {Component} The higher order component.
  */
 
-type HockApplier = (ComposedComponent: ReactClass<any>) => ReactClass<any>;
+declare type HockApplier = (ComposedComponent: ReactClass<any>) => ReactClass<any>;
```
Comment (dancoates) [inline src/typedef/Typedefs.js]: This file needs to be ignored from eslint to stop it throwing errors

File: src/component/table/Table-test.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import test from 'ava';
```
Comment (dancoates) [inline src/component/table/Table-test.js:1]: This file is throwing a bunch of eslint errors. Need to either be fixed or ignored.

Comment (dancoates) [reply src/component/table/Table-test.js:1]: The location of this file is also a little inconsistent. Should we stick to keeping tests in `__test__` folders?

File: src/component/table/Table.jsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        .update(deepReduceOutwards((reduction, item) => {
+            return reduction.push(item);
+        }, List(), [childNodeName]))
+        .skip(1); // The first node is our fake root node.
+
+
+    // Take the schema to create each heading
+    const tableHead: React.Element<any> = schema
+        .map((column, key: number) => <Th key={key} schemaItem={column} />)
+        .toJS();
```
Comment (dancoates) [inline src/component/table/Table.jsx:132]: unnecessary toJS() ?

File: src/component/table/Table.jsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+
+
+
+
+type Schema = ListOrArray<SchemaItem>;
+
+type SchemaItem = {
+    filter: string | (row: Map<any>) => React.Element<any>,
```
Comment (dancoates) [inline src/component/table/Table.jsx]: Not sure about the naming of `filter` it isn't really doing filtering so much as transforming data. It isn't perfect but a possible alternative is what react-virtualized uses [`cellDataGetter`](https://github.com/bvaughn/react-virtualized/blob/master/docs/Column.md)

File: src/component/table/Table.jsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    // but retain their ordering
+    const data = fromJS(props.data)
+        // A fake parent node has to be created so that
+        // the deepReduce can start with a child
+        .update(ii => Map().set(childNodeName, ii))
+        .update(deepReduceOutwards((reduction, item) => {
+            return reduction.push(item);
+        }, List(), [childNodeName]))
+        .skip(1); // The first node is our fake root node.
+
```
Comment (dancoates) [inline src/component/table/Table.jsx]: Not sure that this flattening is the responsibility of the table. Shouldn't it just accept a list and not care what structure your data takes beyond that?

File: src/component/table/Table.jsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+    // Take the schema to create each heading
+    const tableHead: React.Element<any> = schema
```
Comment (dancoates) [inline src/component/table/Table.jsx]: `tableHead` is an array, not a `React.Element<any>` right? Or does `React.Element<any>` encompass arrays?

**Comment by dancoates** [comment]
I've added some comments that are more general discussion than changes.

As well as those it would be good to have a runnable example of all components before they get merged in.

**Comment by dancoates** [comment]
Oh, and I was also thinking that the table has a pretty rigid structure for something that is our most base-level, generic table component. For example there's no way you could have a table head on the left hand side rather than the top. If it is possible to allow for some more customisation then that could be good.

File: src/component/table/Table.jsx (hunk 4 above)
Comment (allanhortle) [reply src/component/table/Table.jsx:132]: Not unnecessary. Things should be plain js at rendertime.

File: src/component/table/Table.jsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+ *
+ * `Table` is a Controlled component that takes a collection of data and a schema. It iterates creating one row per item.
+ *  The schema is used to determine how to render each column.
+ *
+ * ### Content rendering priority order
+ * 1. render function
+ * 2. filter function
+ * 3. filter key accessor
+ *
+ * @param {Object}      [props]
```
Comment (dxinteractive) [inline src/component/table/Table.jsx]: Are all these props really optional?

File: src/component/table/Table.jsx [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+Table.defaultProps = {
+    className: '',
+    childNodeName: 'children',
+    data: List(),
+    modifier: '',
+    schema: List(),
+    rowProps: () => ({}),
+    sortAscending: false,
```
Comment (dxinteractive) [inline src/component/table/Table.jsx]: Are these sort props just left over from something else?

File: src/component/table/Table.jsx [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+function Td(props: TdProps): React.Element<any> {
+    const {row, schemaItem} = props;
+    const {render, filter} = schemaItem.toObject();
+
+    // Content rendering priority order
+    // 1. render function
+    // 2. filter function
+    // 3. filter key accessor
+    const content = (render)
```
Comment (dxinteractive) [inline src/component/table/Table.jsx]: Const should have a flow type

File: src/component/table/Table.jsx [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        .skip(1); // The first node is our fake root node.
+
+
+    // Take the schema to create each heading
+    const tableHead: React.Element<any> = schema
+        .map((column, key: number) => <Th key={key} schemaItem={column} />)
+        .toJS();
+
+    // use `data` then `schema` to make each `row` then `column`
+    const tableBody = data
```
Comment (dxinteractive) [inline src/component/table/Table.jsx]: Const should have flow type

**Review by dxinteractive — CHANGES_REQUESTED** [review]

File: src/component/table/Table-test.js (hunk 3 above)
Comment (allanhortle) [reply src/component/table/Table-test.js:1]: Talking to Damo we decided that the **test** pattern is a bit mystifying. So we might try keeping tests next to their corresponding file just with a `-test.js` extension. 
I've moved the other test.

**Comment by dxinteractive** [comment]
approve

**Review by dxinteractive — APPROVED** [review]

**Comment by dancoates** [comment]
approve

**Pull request merged**